[RFC] Fix/spec system()/jobstart() for Windows

equalsraf opened the pull request:
I've made a mess of things with a previous patch that disabled quoting in Windows  This PR reverts it documents windows behaviour, and adds some functional tests.

The goal of the PR to specify, write test and fix system(list)//jobstart(list) in Windows.
### Background

The `system()` function has two variants
- `system(string)` takes the given string and calls the command `&shell &shellcmdflag <string> (give or tag a flag)
- `system(list)` treats the given list as argv and calls the command `list[0] list[1] ...` bypassing the shell. This variant does not exist in Vim.

the current behavior in UNIX is as follows:
- `system("jobs")` works (in a sh like shell)
- `system(["jobs"])` fails with E902: "jobs" is not an executable, because `jobs` is a shell command
- `system(['python_script'])` works because of the shebang, i.e. `execve()` at work.

In Windows, [uv_spawn()](https://github.com/libuv/libuv/blob/b12624c13693c4d29ca84b3556eadc9e9c0936a4/src/win/process.c) calls [CreateProcess()](https://github.com/libuv/libuv/blob/b12624c13693c4d29ca84b3556eadc9e9c0936a4/src/win/process.c). In particular libuv makes sure to pass in `lpApplicationName`. And verifies this value to some extent.
### system(list)

The current behavior (in this PR) of system() when passed a list is as follows.

Running an .exe or .com file in your $PATH with or without the extension works, e.g.

``` vim
system(["ipconfig"])
system(["ipconfig.exe"])
```

When trying to run a program that is not in your $PATH, you need to specify the path including the extension. For example to run `c:\a.exe`, the following examples WILL NOT WORK

```
system(["c:\\a"])
system(["c:/a"])
```

This **DOES NOT** work, because `dir` is a shell command, not a program

``` vim
system(["dir"])
```

The following works but **SHOULD NOT??**.

``` vim
system(["file.bat"])
" this one does not cause an error, but prints nothing, what is it doing?
system(["file.ps1"])
```

`CreateProcess()` states it cannot run .bat files and `uv_spawn()` states it does not check `PATHEXT`. I have found found this [SO question](http://stackoverflow.com/questions/21553379/createprocess-is-able-to-execute-batch-files-but-documentation-says-the-opposit) that runs into the same issue.

I have tried the following in a minimal program
- CreateProcess('test.bat', ...) works
- CreateProcess('test.ps1', ...) fails

At this point I think this behaviour is undocumented and unreliable. I cannot reproduce at all in Appveyor.

Present uses of `system([...])` in runtime:
- system(['lpr']
- system(['copy', v:fname_in, empty(&printdevice)?'LPT1':&printdevice])
- in pythonx.vim to run `python_prog`
- system(['curl', '-sL', "'", a:url, "'"])
- system(['python', '-c', "'", script, "'", '2>/dev/null'])
- nvim heakth
- system([pyenv, 'prefix']))
- jobstart([...]) in clipboard.vim
- similarly for the ruby provider
### system(string)

The work in fixing system('...') is being handled in #2789/ we have functional tests but those need to be fixed for Windows. A comprehensive test suite for system('...') quoting in the windows shell would be a nice addition to that PR.
### Relation to executable()

Meanwhile`os_can_exe()` used by `executable()` actually checks $PATHEXT.
Current behaviour is:
- `executable('no_such_file')` returns 0
- `executable('file')` returns 1 if file.bat or file.exe,etc exists ($PATHEXT)
- `executable('file.bat')`returns 1
- `executable('file.exe')`returns 1
- In UNIX `executable('filename')` only returns 1 if `filename` is in the $PATH
- In UNIX `executable('./filename')` returns 1 if filename has the executable bit

This could lead to the misconception that, `executable("file.bat")` being `1` implies `system(["file.bat"])` is expected to succeed, which is not true. I have not checked runtime in depth, but most of the uses of executable seem to be meant for the shell, which is the right way.

There are unit tests for `os_can_exe()` but unit tests are not working on windows. There are only couple of tests that call `executable()` (grep, terminal buffer). Well handle that in a followup PR.
### Notes
- The `printargs` helper has to be in C, a bat file is not an option, it would cause the shelll to be used, and we want avoid that.

equalsraf committed:
Add functional tests for system([...])

Commit:
Revert "Windows: disable libuv argument escaping for uv_spawn #3305"

This reverts commit dc9652e68de163290abee880a74bf1727c715a1e.

Disabling the quoting was does not solve the problem in general, and
we would end up having to handle the quoting ourselves.

Commit:
Remove old docs for system()

The docs already reference jobstart(), and these two paragraphs
no longer apply.

Commit:
Windows: Update jobstart docs

equalsraf commented:
errr, why would `executable('README.md')`return 1? Also happens in gVim in Windows.

equalsraf commented:
system(['python.exe', '-c', 'import sys']) fails with syntax error

equalsraf commented:
The python3 provider seems to be working, but FYI the ruby provider, https://github.com/neovim/neovim/blob/71b3e20d0fba20d4957c1949352bde1744a0a947/runtime/autoload/provider/ruby.vim#L19, is a batch file.

equalsraf committed:
Appveyor: Enable Python provider tests

Setup python2/3 and install the Neovim client in Appveyor to
enable the python tests.

We are using the Python instalation provided by Appveyor, because
dependencies pyuv, greenlet have issues compiling with MinGW-w64.
And this way we avoid building those too.

We create a copy of python.exe as python3.exe to workaround the lack
of version specific alias for python, that prevents the python provider
from finding python3.

jamessan commented:
@jszakmeister QB seems to be [stuck] on this PR.  Any chance you can unwedge it?

[stuck]: http://neovim-qb.szakmeister.net/build/8220

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -6817,20 +6826,10 @@ synstack({lnum}, {col})					*synstack()*
-		in a way |writefile()| does with {binary} set to "b" (i.e.
-		with a newline between each list item with newlines inside
-		list items converted to NULs).  
-		Pipes are not used.
```
We still support `{input}`, so why remove this whole chunk?

equalsraf replied:
It seems I removed too much. It should have been replaced with a smaller paragraph, from the comment in https://github.com/neovim/neovim/pull/3812/files#r47184131

justinmk commented on test/functional/shell/viml_system_spec.lua:
```diff
@@ -195,15 +231,10 @@ describe('system()', function()
-      eq('* $NOTHING ~/file',
-         eval("system(['echo', '-n', '*', '$NOTHING', '~/file'])"))
-    end)
-  end)
```
Ha, I forgot we supported this (doesn't seem to be documented). It's useful, like a synchronous `jobstart()`. Though I guess it violates the analogy of `system()` from C stdlib.

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -6826,19 +6835,14 @@ synstack({lnum}, {col})					*synstack()*
 
 system({cmd} [, {input}])				*system()* *E677*
-		Get the output of the shell command {cmd} as a |string|.  {cmd}
+		Get the output of the command {cmd} as a |string|.  {cmd}
```
We should document "`{cmd}` as List" behavior.

equalsraf replied:
I assumed the idea here was to point towards the docs for jobstart() rather than repeating the same text.

justinmk commented:
> Present uses of `system([...])` in runtime:
>
> system(['python', '-c', "'", script, "'", '2>/dev/null'])

That looks like a runtime bug, stderr redirection won't work if `system({List})` does not invoke a shell.

> This could lead to the misconception that, executable("file.bat") being 1 implies system(["file.bat"]) is expected to succeed

Not much we can do about that, but that's why it's risky to change the semantics of `system()`. Too late now I guess, but we can at least note this in the docs. ("system({List}) does not invoke shell, therefore executable() result is not valid for shell builtins in determining whether system({List}) can invoke it")

> `findstr x*` was the closest I found to cat - however it leaves a trailing newline

powershell has `cat`. If the utf16 streams can be handled, I think we should start using powershell as much as possible.

equalsraf commented:
> Not much we can do about that, but that's why it's risky to change the semantics of system().

Its ok as long as its documented, since the original semantics are still available using `system(string)`, worst case scenario there will be some places where we call system(str) and others where we call system(list) depending on the platform.

Also this is the correct semantics if we consider that &shell might not be what the caller expects. If the caller wants to run a batch file, it really should call `system(['cmd', '/c', 'batch ...'])`. It certainly seems better than temporarily changing &shell and similar options just to run a command.

> That looks like a runtime bug, stderr redirection won't work if system({List}) does not invoke a shell.

https://github.com/neovim/neovim/blob/master/runtime/autoload/health/provider.vim#L53 it is our own script being called, so we can probably avoid stderr from within it.

But I didn't get a chance to check CheckHealth in windows. It seems to be failing to run for some other reason.

equalsraf commented:
Ping. Travis started failing since the rebase, but seems unrelated. Can't figure out the issue with QB.

jamessan commented:
> Travis started failing since the rebase, but seems unrelated.

Yeah, the tab drag test has been pretty flaky lately.  I've restarted that build, but if that's the only failure, it's fine.  The OS X test failures are during 

> Can't figure out the issue with QB.

All the QB failures are in this test:

> 18:08:17,517 INFO  - not ok 1432 - system() command passed as a list quotes arguments correctly
18:08:17,517 INFO  - # test/functional/shell/viml_system_spec.lua @ 42
18:08:17,517 INFO  - # Failure message: test/functional/shell/viml_system_spec.lua:44: Expected objects to be the same.
18:08:17,517 INFO  - # Passed in:
18:08:17,517 INFO  - # (string) ''
18:08:17,517 INFO  - # Expected:
18:08:17,517 INFO  - # (string) 'arg1=1;arg2=2 "3;'
18:08:17,517 INFO  - # stack traceback:
18:08:17,517 INFO  - #     test/functional/shell/viml_system_spec.lua:44: in function <test/functional/shell/viml_system_spec.lua:42>
18:08:17,517 INFO  - #

jszakmeister commented:
> @jszakmeister QB seems to be stuck on this PR. Any chance you can unwedge it?

@jamessan FYI: it's better to email me directly rather than ping me through GitHub notifications.  I find that the notifications are getting lost in the noise. :-(  My job keeps me away from the Internet quite a lot, so it may take me a bit to get around to it.  I'm hoping to find a better way to run the test (such as under a process group) so that I can implement a better way of killing all the child processes.  QB's technique is racy at best. :-(

jamessan commented:
> FYI: it's better to email me directly rather than ping me through GitHub notifications.

Noted.  I thought you had already handled it, as QB has been processing builds again for a couple weeks now.

jszakmeister commented:
> Noted. I thought you had already handled it, as QB has been processing builds again for a couple weeks now.

Yeah, I noticed it.  It was more of a "in the future" thing.  I hate leaving you guys hanging. :-)

justinmk commented:
@equalsraf Any problems left with this?

equalsraf commented:
@justinmk the new parseargs tests were failing on QB, with an empty string in multiple systens (freebsd and Mac I think).

wsdjeg commented:
@equalsraf I just test #5280, everything works well, but it make some func in tagbar.vim  can not work.
and here is the test func. in tagbar.vim the code is [link](https://github.com/majutsushi/tagbar/blob/master/autoload/tagbar.vim#L3694-L3737)
```vim
fu! Test(ctags_cmd)
    let s:debug = 0
    if &shell =~# 'fish$'
        " Reset shell since fish isn't really compatible
        let shell_save = &shell
        set shell=sh
    endif

    if exists('+shellslash')
        let shellslash_save = &shellslash
        set noshellslash
    endif

    if &shell =~ 'cmd\.exe'
        let shellxquote_save = &shellxquote
        set shellxquote=\"
        let shellcmdflag_save = &shellcmdflag
        set shellcmdflag=/s\ /c
    endif

    if s:debug
        silent 5verbose let ctags_output = system(a:ctags_cmd)
        call s:debug(v:statusmsg)
        redraw!
    else
        silent let ctags_output = system(a:ctags_cmd)
    endif

    if &shell =~ 'cmd\.exe'
        let &shellxquote  = shellxquote_save
        let &shellcmdflag = shellcmdflag_save
    endif

    if exists('+shellslash')
        let &shellslash = shellslash_save
    endif

    if exists('shell_save')
        let &shell = shell_save
    endif

    return ctags_output
endf
```
then in gvim `:echo Test('ctags --version')`, you will see
```
Universal Ctags 0.0.0, Copyright (C) 2015 Universal Ctags Team
Universal Ctags is derived from Exuberant Ctags.
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
  Compiled: Sep  9 2016, 04:41:00
  URL: https://ctags.io/
  Optional compiled features: +win32, +wildcards, +regex, +internal-sort, +multibyte
```
but it failed with #5280 , with #5280 you will see:
![tagbar-error](https://cloud.githubusercontent.com/assets/13142418/18958138/3839f52c-8694-11e6-8375-3a63bfe31c20.png)

BTW, it this Test() func works well in neovim-master

equalsraf commented:
@wsdjeg I can't reproduce that last issue. `:echo Test('ctags --version')`

```
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
Compiled: Sep  2 2015, 09:00:12
Addresses: <dhiebert@users.sourceforge.net>, http://ctags.sourceforge.net
Optional compiled features: +win32, +regex, +internal-sort
```

Where did you get you ctags?

In you screenshot I don't see any error, it seems like the output  of the command is messed up.

wsdjeg commented:
I am using [ctags](https://github.com/universal-ctags/ctags), and download from [link](https://ci.appveyor.com/project/masatake/ctags/branch/master/job/dpl8ohhi1i4mm4rx/artifacts), and my neovim is downloaded from [link](https://ci.appveyor.com/project/neovim/neovim/build/8545/job/2vqohfk2shm2ygwb/artifacts)

equalsraf committed:
system([]) must set v:shell_error to -1 on error

Calling system(['no-such file']) failed to set v:shell_error when
the executable does not exist.

equalsraf commented:
QB log

```
17:47:50,248 INFO  - not ok 1432 - system() command passed as a list quotes arguments correctly
17:47:50,248 INFO  - # test/functional/shell/viml_system_spec.lua @ 42
17:47:50,248 INFO  - # Failure message: test/functional/shell/viml_system_spec.lua:44: Expected objects to be the same.
17:47:50,248 INFO  - # Passed in:
17:47:50,248 INFO  - # (string) ''
17:47:50,248 INFO  - # Expected:
17:47:50,248 INFO  - # (string) 'arg1=1;arg2=2 "3;'
17:47:50,248 INFO  - # stack traceback:
17:47:50,248 INFO  - # 	test/functional/shell/viml_system_spec.lua:44: in function <test/functional/shell/viml_system_spec.lua:42>
```

system([]) can fail while returning an empty string. For example this test succeeds in my Linux system.

```lua
  it('XXXXX', function()
    local out = helpers.call('system', {'XXXXX', '1', '2'})
    eq(0, eval('v:shell_error'))
    eq('', out)
  end)
```

Added a patch to set v:shell_error to -1, but I'm still not getting v:errmsg

jamessan commented:
> Added a patch to set v:shell_error to -1, but I'm still not getting v:errmsg

What are you expecting `v:errmsg` to be set?  That should only occur for errors produced by nvim, not due to an external command failing to run.

equalsraf commented:
I was expecting E902: \"%s\" is not an executable

Commit:
debug QB

equalsraf commented:
I'm out of ideas on this one.

I seems QB is not building the printargs target at all, but it should be built wih  shell-test.

equalsraf commented:
Yup the command in configure-neovim-and-build-nvim is

```
mkdir -p build/${params.get("buildType")} && cd build/${params.get("buildType")} && cmake -G "Unix Makefiles" -DBUSTED_OUTPUT_TYPE=TAP -DCMAKE_BUILD_TYPE=${params.get("buildType")} -DTRAVIS_CI_BUILD=ON ../.. && ${node.getAttribute("make", "make")} VERBOSE=1 nvim nvim-test unittest-headers tty-test shell-test
```

thats why printargs is missing

equalsraf commented:
@wsdjeg do you still have the same issue with ctags? 

Can you test with `:echo Test('bin\printargs --version')`

**printargs** is a test binary built with nvim.exe.

justinmk commented:
@jszakmeister does a QB script need to be updated to fix https://github.com/neovim/neovim/pull/5280#issuecomment-251098799 ?

wsdjeg commented:
@equalsraf I just downlowd nvim from https://ci.appveyor.com/project/neovim/neovim-1kmqj/build/8580/job/lx3d1he04gydgnhc/artifacts,  but I can not run nvim-qt, it show me 
![qq 20161005200300](https://cloud.githubusercontent.com/assets/13142418/19112306/c0d46c4a-8b36-11e6-8e49-feefa9727de7.png)

equalsraf commented:
@wsdjeg Is it the latest version of nvim-qt(from the master branch)?

wsdjeg commented:
@equalsraf yeah, I download it from the release page, and the url is `https://github.com/equalsraf/neovim-qt/releases/download/v0.2.3/neovim-qt.zip`

equalsraf commented:
@wsdjeg I've tagged a new release as **staging**, can you try that one?

wsdjeg commented:
@equalsraf just same as before,
![qq 20161005214851](https://cloud.githubusercontent.com/assets/13142418/19115806/ddb068f0-8b45-11e6-99ec-aa892d638274.png)

as shown in Test()
```
if &shell =~ 'cmd\.exe'
        let shellxquote_save = &shellxquote
        set shellxquote=\"
        let shellcmdflag_save = &shellcmdflag
        set shellcmdflag=/s\ /c
endif
```

my shell is `cmd.exe`, so the value of shellxquote and shellcmdflag has been changed.
now I use my fork [tagbar.vim](https://github.com/wsdjeg/tagbar) it works well in nvim-qt in win7.

justinmk commented on src/nvim/eval.c:
```diff
@@ -16043,6 +16043,7 @@ static void get_system_output_as_rettv(typval_T *argvars, typval_T *rettv,
   // get shell command to execute
   char **argv = tv_to_argv(&argvars[0], NULL);
   if (!argv) {
+    set_vim_var_nr(VV_SHELL_ERROR, (long)-1);
```
In #5910 I changed this to only set `v:shell_error` if `tv_to_argv()` returns `executable==false`. The other failures are parameter validation, which should _not_ set `v:shell_error`.

justinmk commented:
Continued at https://github.com/neovim/neovim/pull/5910. Thank you @equalsraf
